libbpf-tools/cpudist.c

```diff
@@ -0,0 +1,314 @@
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static int bump_memlock_rlimit(void)
+{
+	struct rlimit rlim_new = {
+		.rlim_cur	= RLIM_INFINITY,
+		.rlim_max	= RLIM_INFINITY,
+	};
+
+	return setrlimit(RLIMIT_MEMLOCK, &rlim_new);
+}
+
+static int get_pid_max(void)
+{
+	int pid_max;
+	FILE *f;
+
+	f = fopen("/proc/sys/kernel/pid_max", "r");
+	if (!f)
+		return -1;
+	if (fscanf(f, "%d\n", &pid_max) != 1)
+		pid_max = -1;
+	fclose(f);
+	return pid_max;
+}
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+static void pid_to_comm(pid_t pid, char *comm)
+{
+	char fname[64];
+	FILE *f;
+
+	snprintf(fname, sizeof(fname), "/proc/%d/comm", pid);
+	f = fopen(fname, "r");
+	if (!f || fscanf(f, "%s\n", comm) < 0)
+		*comm = '\0';
+	if (f)
+		fclose(f);
+}
+
+static int print_log2_hists(int fd)
+{
+	char *units = env.milliseconds ? "msecs" : "usecs";
+	__u32 lookup_key = -2, next_key;
+	struct hist hist;
+	char comm[16];
+	int err;
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_lookup_elem(fd, &next_key, &hist);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return -1;
+		}
+		pid_to_comm(next_key, comm);
+		if (env.per_process)
+			printf("\npid = %d %s\n", next_key, comm);
+		if (env.per_thread)
+			printf("\ntid = %d %s\n", next_key, comm);
+		print_log2_hist(hist.slots, MAX_SLOTS, units);
+		lookup_key = next_key;
+	}
+
+	lookup_key = -2;
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_delete_elem(fd, &next_key);
+		if (err < 0) {
+			fprintf(stderr, "failed to cleanup hist : %d\n", err);
+			return -1;
+		}
+		lookup_key = next_key;
+	}
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct cpudist_bpf *obj;
+	int pid_max, fd, err;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = cpudist_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF ojbect\n");
+		return 1;
+	}
+
+	/* initialize global data (filtering options) */
+	obj->rodata->targ_per_process = env.per_process;
+	obj->rodata->targ_per_thread = env.per_thread;
+	obj->rodata->targ_ms = env.milliseconds;
+	obj->rodata->targ_offcpu = env.offcpu;
+	obj->rodata->targ_tgid = env.pid;
+
+	pid_max = get_pid_max();
+	if (pid_max < 0) {
+		fprintf(stderr, "failed to get pid_max\n");
+		return 1;
+	}
+
+	bpf_map__resize(obj->maps.start, pid_max);
```

Comment:
@anakryiko  Thanks for these details. But I met a problem with using `ARRAY`:

I changed two place:

1. `__uint(type, BPF_MAP_TYPE_HASH)` to `__uint(type, BPF_MAP_TYPE_ARRAY)`
2. 
```c
    if (!tsp || ts < *tsp)
        return;
```
to

 ```c
    if (!tsp || *tsp == 0 || ts < *tsp)
        return;
```

Then got some wrong tsp cause loc >= MAX_SLOTS.  


```shell
           cat-3560  [030] d... 5145650.191445: 0: loc: 35, *ts: 5145651753477337, *tsp: 5145651753477473
           <...>-2790  [024] d... 5145650.321661: 0: loc: 35, *ts: 5145651883695616, *tsp: 5145651883695679
   node_exporter-4345  [014] d... 5145650.324224: 0: loc: 35, *ts: 5145651886258518, *tsp: 5145651886258528
             cat-3560  [027] d... 5145650.348537: 0: loc: 35, *ts: 5145651910571538, *tsp: 5145651910571657
  systemd-logind-2250  [021] d... 5145651.015160: 0: loc: 35, *ts: 5145652577205888, *tsp: 5145652577205942
             cat-3560  [017] d... 5145651.015339: 0: loc: 35, *ts: 5145652577384536, *tsp: 5145652577384614
        libvirtd-3666  [036] d... 5145651.019308: 0: loc: 35, *ts: 5145652581353616, *tsp: 5145652581353662
             cat-3560  [017] d... 5145651.295419: 0: loc: 35, *ts: 5145652857469206, *tsp: 5145652857469271
             cat-3560  [017] d... 5145651.439423: 0: loc: 35, *ts: 5145653001474939, *tsp: 5145653001474999
   rs:main Q:Reg-5835  [025] d... 5145651.762465: 0: loc: 35, *ts: 5145653324521559, *tsp: 5145653324521681
           snapd-29958 [038] d... 5145651.764726: 0: loc: 35, *ts: 5145653326784044, *tsp: 5145653326784101
     dbus-daemon-2254  [021] d... 5145651.765854: 0: loc: 35, *ts: 5145653327912088, *tsp: 5145653327912402
   systemd-udevd-3645  [029] d... 5145651.766828: 0: loc: 35, *ts: 5145653328883935, *tsp: 5145653328884135
   systemd-udevd-3647  [031] d... 5145651.766852: 0: loc: 35, *ts: 5145653328908609, *tsp: 5145653328908646
   systemd-udevd-3648  [033] d... 5145651.933379: 0: loc: 35, *ts: 5145653495438366, *tsp: 5145653495438811
             cat-3560  [019] d... 5145651.963414: 0: loc: 35, *ts: 5145653525475638, *tsp: 5145653525475745
```

I'm confusing about this.

Reply:
sorry, can you point out specifically what's wrong there?

Reply:
> sorry, can you point out specifically what's wrong there?

@anakryiko Sorry, I didn't describe it clearly. I found when I use HASHMAP,  no slot will >= MAX_SLOTS, but after I changed to ARRAY, it always have slot >= MAX_SLOTS. 

I ran with `cpudist -L 1` (iter hists 1s interval), I found only tid == 0 will produce slot >= MAX_SLOTS when map type is ARRAY.

```
tid = 0 swapper/36
               usecs                         : count     distribution
                   0 -> 1                    : 28       |***                 |
                   2 -> 3                    : 31       |****                |
                   4 -> 7                    : 154      |********************|
                   8 -> 15                   : 68       |********            |
                  16 -> 31                   : 58       |*******             |
                  32 -> 63                   : 100      |************        |
                  64 -> 127                  : 17       |**                  |
                 128 -> 255                  : 11       |*                   |
                 256 -> 511                  : 9        |*                   |
                 512 -> 1023                 : 16       |**                  |
                1024 -> 2047                 : 25       |***                 |
                2048 -> 4095                 : 18       |**                  |
                4096 -> 8191                 : 14       |*                   |
                8192 -> 16383                : 9        |*                   |
               16384 -> 32767                : 3        |                    |
               32768 -> 65535                : 0        |                    |
               65536 -> 131071               : 0        |                    |
              131072 -> 262143               : 0        |                    |
              262144 -> 524287               : 0        |                    |
              524288 -> 1048575              : 0        |                    |
             1048576 -> 2097151              : 0        |                    |
             2097152 -> 4194303              : 0        |                    |
             4194304 -> 8388607              : 0        |                    |
             8388608 -> 16777215             : 0        |                    |
            16777216 -> 33554431             : 0        |                    |
            33554432 -> 67108863             : 0        |                    |
            67108864 -> 134217727            : 0        |                    |
           134217728 -> 268435455            : 0        |                    |
           268435456 -> 536870911            : 0        |                    |
           536870912 -> 1073741823           : 0        |                    |
          1073741824 -> 2147483647           : 0        |                    |
          2147483648 -> 4294967295           : 0        |                    |
          4294967296 -> 8589934591           : 0        |                    |
          8589934592 -> 17179869183          : 0        |                    |
         17179869184 -> 34359738367          : 5        |                    |
```

Reply:
it's probably because bpf_map_lookup_elem() for ARRAY never returns NULL and check doesn't account for that. So your algorithm will assume that tid=0 was already timestamped or something. Either way, let's stick to HASHMAP for now and simplify our lives.

Reply:
Actually, just running your code locally (with HASHMAP), I still got a really large value. Probably also from tid=0. I wonder if it makes sense to just filter that out? What's the behavior of BCC-based cpuhist in such case?

Reply:
> Actually, just running your code locally (with HASHMAP), I still got a really large value. Probably also from tid=0. I wonder if it makes sense to just filter that out? What's the behavior of BCC-based cpuhist in such case?

@anakryiko with HASHMAP, I ran`crudest -p 0`  for 15 minutes, after ctrl-c, the result is:

```
Tracing on-CPU time... Hit Ctrl-C to end.
^C
     usecs               : count    distribution
         0 -> 1          : 25823    |*****                                   |
         2 -> 3          : 30616    |******                                  |
         4 -> 7          : 49292    |**********                              |
         8 -> 15         : 88110    |******************                      |
        16 -> 31         : 123735   |**************************              |
        32 -> 63         : 122768   |*************************               |
        64 -> 127        : 189345   |****************************************|
       128 -> 255        : 34011    |*******                                 |
       256 -> 511        : 26639    |*****                                   |
       512 -> 1023       : 18085    |***                                     |
      1024 -> 2047       : 18400    |***                                     |
      2048 -> 4095       : 54752    |***********                             |
      4096 -> 8191       : 25210    |*****                                   |
      8192 -> 16383      : 9261     |*                                       |
     16384 -> 32767      : 12301    |**                                      |
     32768 -> 65535      : 6091     |*                                       |
     65536 -> 131071     : 446      |                                        |
    131072 -> 262143     : 31       |                                        |
```

No large value found. I tested on a physical machine:

```
CPU(s):              40
On-line CPU(s) list: 0-39
Thread(s) per core:  2
Core(s) per socket:  10
Socket(s):           2
NUMA node(s):        2
```

The bcc version for almost half an hour.

```
➜  tools git:(libbpf-tools) ✗ ./cpudist.py -p 0
Tracing on-CPU time... Hit Ctrl-C to end.
^C
     usecs               : count     distribution
         0 -> 1          : 64033    |******                                  |
         2 -> 3          : 59896    |******                                  |
         4 -> 7          : 96904    |**********                              |
         8 -> 15         : 176765   |******************                      |
        16 -> 31         : 247373   |*************************               |
        32 -> 63         : 248718   |*************************               |
        64 -> 127        : 387502   |****************************************|
       128 -> 255        : 68589    |*******                                 |
       256 -> 511        : 53117    |*****                                   |
       512 -> 1023       : 37782    |***                                     |
      1024 -> 2047       : 37948    |***                                     |
      2048 -> 4095       : 109874   |***********                             |
      4096 -> 8191       : 50249    |*****                                   |
      8192 -> 16383      : 18970    |*                                       |
     16384 -> 32767      : 24702    |**                                      |
     32768 -> 65535      : 13212    |*                                       |
     65536 -> 131071     : 1151     |                                        |
    131072 -> 262143     : 67       |                                        |

```

Reply:
I mean, this will depend on specific machine's workload, but sure, let's see if that becomes a problem.